src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java

```diff
@@ -48,81 +51,133 @@
+        Map<InetAddressAndPort, String> datacenterMap = peers.stream()
+                                                             .collect(Collectors.toMap(k -> k, getDatacenterSource));
+        Function<InetAddressAndPort, String> getDatacenter = datacenterMap::get;
 
-        long startNanos = System.nanoTime();
+        Map<String, Set<InetAddressAndPort>> peersByDc = peers.stream()
+                                                              .collect(Collectors.groupingBy(getDatacenter,
+                                                                                             Collectors.toSet()));
+
+        if (!blockForRemoteDcs)
+        {
+            peersByDc.keySet().retainAll(Collections.singleton(localDc));
+            logger.info("Blocking coordination until only a single peer is DOWN in the local datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
+        else
+        {
+            logger.info("Blocking coordination until only a single peer is DOWN in each datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
 
         AckMap acks = new AckMap(3);
-        int target = (int) ((targetPercent / 100.0) * peers.size());
-        CountDownLatch latch = new CountDownLatch(target);
+        Map<String, CountDownLatch> latchMap = new HashMap<>(peersByDc.size());
+        for (Map.Entry<String, Set<InetAddressAndPort>> entry: peersByDc.entrySet())
+        {
+            latchMap.put(entry.getKey(), new CountDownLatch(Math.max(entry.getValue().size() - 1, 0)));
+        }
+
+        long startNanos = System.nanoTime();
 
         // set up a listener to react to new nodes becoming alive (in gossip), and account for all the nodes that are already alive
-        Set<InetAddressAndPort> alivePeers = Sets.newSetFromMap(new ConcurrentHashMap<>());
-        AliveListener listener = new AliveListener(alivePeers, latch, acks);
+        Set<InetAddressAndPort> alivePeers = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        AliveListener listener = new AliveListener(alivePeers, latchMap, acks, getDatacenter);
         Gossiper.instance.register(listener);
 
         // send out a ping message to open up the non-gossip connections
-        sendPingMessages(peers, latch, acks);
+        sendPingMessages(peers, latchMap, acks, getDatacenter);
 
         for (InetAddressAndPort peer : peers)
+        {
             if (Gossiper.instance.isAlive(peer) && alivePeers.add(peer) && acks.incrementAndCheck(peer))
-                latch.countDown();
+            {
+                String datacenter = getDatacenter.apply(peer);
+                if (latchMap.containsKey(datacenter))
+                    latchMap.get(datacenter).countDown();
+            }
+        }
+
+        boolean succeeded = Uninterruptibles.awaitUninterruptibly(latchMap.get(localDc), timeoutNanos, TimeUnit.NANOSECONDS);
```

Comment:
I think the loop is requires because this main method can race with the callbacks that are triggered as part of {{sendPingMessages}}. There is no point in the {{containsKey}} though so I just refactored back to what was there before.